src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,495 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
```

Comment:
What if the function had such a signature?
```cpp
void
CQueryMutators::GroupingListContainsPrimaryKey(SContextFixGroupDependentTargets *context, List *new_list)
```
And it will add context->m_gc to new_list if necessary.

Reply:
I do not like this idea.

1. Intention of ```GroupingListContainsPrimaryKey``` is to perform some check.
You do not expect that this or similar function will perform some modifications (otherwise it should be renamed).
2. Now this function is a utility with quite straightforward purpose. It has a potential to be re-used in other places if needed. And I want to keep it this way.
	2.1 Adding more responcibilities to this function will make it too specialized and forbid possible re-use of this function in other places.
	2.2 Changing the function param to context will also make strong coupling with the mutator, and will complicate potential re-use of this function.

3. It would be better to keep the code that modifies the tree in the mutator itself, because it is mutator's responcibility to change the tree. And the mutator's code is not so complicated to move it to inner functions.

Having said all the above, I would like to keep this part of the code as is.

Reply:
> 3\. t would be better to keep the code that modifies the tree in the mutator itself, because it is mutator's responcibility to change the tree. And the mutator's code is not so complicated to move it to inner functions.

I do not suggest changing the tree in this function, but only the new_list passed as an argument.

Reply:
> 2\. Now this function is a utility with quite straightforward purpose. It has a potential to be re-used in other places if needed. And I want to keep it this way.

Where can this function be reused with such arguments?

Reply:
> Where can this function be reused with such arguments?

I look at it from a bit different angle:  a function ```GroupingListContainsPrimaryKey``` is much more likely to be used elsewhere than a function ```AddSomethingToTheListIfGroupingListContainsPrimaryKey```.

Plus now if one wants to understand how the query is transformed by AddMissingGroupClauseMutator, he do not need to dig into any nested functions - it should be pretty clear what GroupingListContainsPrimaryKey  does and that it doesn't alter the query structure.

If we change it in the proposed way, it will be harder to read the code from my point of view - as you need to go back and forth between these functions to compose the full picture.

Reply:
@RekGRpth , do you accept the arguments above?